ui/src/app/tags/components/TagsHeader/AddTagForm/AddTagForm.test.tsx

```diff
@@ -0,0 +1,104 @@
+import { fireEvent, render, screen, waitFor } from "@testing-library/react";
+import userEvent from "@testing-library/user-event";
+import { createMemoryHistory } from "history";
+import { Provider } from "react-redux";
+import { MemoryRouter, Route, Router } from "react-router-dom";
+import configureStore from "redux-mock-store";
+
+import AddTagForm from "./AddTagForm";
+import { Label } from "./AddTagFormFields";
+
+import * as baseHooks from "app/base/hooks/base";
+import type { RootState } from "app/store/root/types";
+import { actions as tagActions } from "app/store/tag";
+import tagsURLs from "app/tags/urls";
+import {
+  tag as tagFactory,
+  rootState as rootStateFactory,
+  tagState as tagStateFactory,
+} from "testing/factories";
+
+const mockStore = configureStore();
+
+let state: RootState;
+
+beforeEach(() => {
+  state = rootStateFactory({
+    tag: tagStateFactory(),
+  });
+  jest
+    .spyOn(baseHooks, "useCycled")
+    .mockImplementation(() => [false, () => null]);
+});
+
+afterEach(() => {
+  jest.restoreAllMocks();
+});
+
+it("dispatches an action to create a tagrange", async () => {
+  const store = mockStore(state);
+  render(
+    <Provider store={store}>
+      <MemoryRouter initialEntries={[{ pathname: "/tags", key: "testKey" }]}>
+        <AddTagForm onClose={jest.fn()} />
+      </MemoryRouter>
+    </Provider>
+  );
+  userEvent.type(screen.getByRole("textbox", { name: Label.Name }), "name1");
+  userEvent.type(
+    screen.getByRole("textbox", { name: Label.Definition }),
+    "definition1"
+  );
+  userEvent.type(
+    screen.getByRole("textbox", { name: Label.Comment }),
+    "comment1"
+  );
+  userEvent.type(
+    screen.getByRole("textbox", { name: Label.KernelOptions }),
+    "options1"
+  );
+  fireEvent.submit(screen.getByRole("form"));
```

Comment:
It would be better to write in this way that resembles real user interaction (either clicking a submit button, or  `userEvent.type(input, "{enter}");`.

Reply:
This is something I've been curious about. Doesn't testing for one particular interaction actually limit (or further dictate) the expected interactions with the form, when in reality the user can submit the form with pressing enter in the form or clicking the submit (I also don't know if there are some screen readers/mobile keyboards etc. can call `onSubmit()` directly).

It seems like that (without testing every possible interaction) testing the form using `.submit()` captures the broadest scope of interactions as opposed to having a test that would pass with a `<button onClick=...`

Reply:
> This is something I've been curious about. Doesn't testing for one particular interaction actually limit (or further dictate) the expected interactions with the form, when in reality the user can submit the form with pressing enter in the form or clicking the submit (I also don't know if there are some screen readers/mobile keyboards etc. can call `onSubmit()` directly).

Very good point. It is a compromise, for sure. The way I think about it is: test from a real user's perspective as much as you can.

A user normally can never fire a submit event directly and it's better to focus on real interactions. In my experience, testing the enter button press is often the best choice (clicking the button is more likely to be caught with manual testing).

But you did spark my interest, I'll dig into this a bit more.

> It seems like that (without testing every possible interaction) testing the form using `.submit()` captures the broadest scope of interactions as opposed to having a test that would pass with a `<button onClick=...`

It would be great if it worked that way in practice, but in reality (AFAIK) it's possible that `fireEvent.submit` would trigger the form submit correctly when neither submit button click nor pressing Enter works.

Reply:
So maybe the approach here should be that the FormikForm component should have tests for all three interactions, while the components that consume FormikForm can use any of the methods. We don't need to test the internals of the component each time we consume it, but we also get the benefit of having tests for each way the user can interact with the form.

I've filed https://github.com/canonical-web-and-design/maas-ui/issues/3688 so we can continue the discussion there.